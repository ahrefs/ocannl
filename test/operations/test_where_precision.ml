open Base
module Tensor = Ocannl.Tensor
module Train = Ocannl.Train
module NTDSL = Ocannl.Operation.NTDSL
module Tn = Ir.Tnode

let () =
  Tensor.unsafe_reinitialize ();
  let module Backend = (val Backends.fresh_backend ()) in
  (* Test heterogeneous precision in where operation using %cd syntax *)
  let%cd condition = { cond } in
  let%cd then_val = { a } in
  let%cd else_val = { b } in
  let%cd result = where condition then_val else_val in

  (* Set different precisions for each argument *)
  Tn.update_prec condition.value Ir.Ops.byte;
  (* condition uses byte precision *)
  Tn.update_prec then_val.value Ir.Ops.half;
  (* then branch is half *)
  Tn.update_prec else_val.value Ir.Ops.bfloat16;
  (* else branch is bfloat16 *)
  Tn.update_prec result.value Ir.Ops.single;

  (* result is single *)

  (* Initialize on host (rather than on device). *)
  Tn.set_values cond.value [| 0.0 |];
  Tn.set_values a.value [| 1.0 |];
  Tn.set_values b.value [| 2.0 |];

  (* Set up values and run computation *)
  ignore (Train.forward_once (module Backend) result);
  Train.printf_tree result
