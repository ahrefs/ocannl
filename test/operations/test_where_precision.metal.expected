#include <metal_stdlib>
using namespace metal;
kernel void where_fwd(
    device bfloat *b_else_val [[buffer(0)]],
    device float *where [[buffer(1)]],
    device half *a_then_val [[buffer(2)]],
    device uchar *cond_condition [[buffer(3)]],
    uint3 gid [[threadgroup_position_in_grid]],
    uint3 lid [[thread_position_in_threadgroup]]) {

  /* Local declarations and initialization. */

  /* Main logic. */
  /* where fwd */
  where[0] = select((float)(b_else_val[0]), (float)(a_then_val[0]), cond_condition[0]);
  /* end */
}